p4-16/psa/PSA.mdk

```diff
@@ -523,34 +523,71 @@ extern Checksum<W> {
 
 ## Counters
 
-Counters are a simple mechanism for keeping statistics about the
-packets that trigger a table in a Match Action unit.
+Counters are a mechanism for keeping statistics.  The control plane
+can read counter values.  A P4 program cannot read counter values,
+only update them.  If you wish to implement a feature involving
+sequence numbers, for example, use Registers instead (Section
+[#sec-registers]).
+
+Direct counters are counters associated with a particular P4 table,
+and are implemented by the extern `DirectCounter`.  There are also
+'indirect' counters, which are implemented by the extern `Counter`.
+The primary differences between direct counters and indirect counters
+are:
+
+- Number of independently updatable counter values:
+  - A single instantiation of a direct counter always contains as many
+    independent counter values as the number of entries in the table
+    with which it is associated.
+  - You must specify the number of independent counter values for an
+    indirect counter when instantiating it.  This number of counters
+    need not be the same as the size of any P4 table.
+- Where counter updates are allowed in the P4 program:
+  - For a direct counter, you may only invoke its `count` method from
+    inside the actions of the table with which it is associated, and
+    this always updates the counter value associated with the matching
+    table entry.
+  - For an indirect counter, you may invoke its `count` method
+    anywhere in the P4 program where extern object method invocations
+    are permitted (e.g. inside actions, or directly inside a control's
+    `apply` block), and every such invocation must specify the index
+    of the counter value to be updated.
```

Comment:
@jafingerhut 

I think I am still struggling to internalize the scope and the goals of the PSA project and the comment above is the indication of this struggle. We had some discussion, but I do not believe we codified that. 

We are working on *Portable* Switch Architecture, so I guess that portability is one of the the goals (if not the main one) :) The question is how do we define portability and, subsequently PSA-compliance?

Usually, we can define portability as the ability of the program to produce the same results regardless of the target. As a consequence, we can check the compliance by ensuring that all required objects/primitives/functions have been defined and pass the functional tests. The ability of the program to be compiled is not questioned or it is easily quantifiable (e.g. C specification states that the compiler should accept identifiers that are at least 32 characters long, that it should accept programs that have at least 50 levels of statement nesting, etc.) 

In P4 world, the simple ability of the program to compile is actually not given, at least not for the ASIC targets (as opposed to SW models, NPUs, FPGAs, etc.). Therefore, the PSA implementation for a given target can have all the objects correctly defined and yet a program, utilizing PSA architecture and developed on a different target might not at all be compilable and not only because of the lack of resources, but because it violates other kinds of restrictions (can't increment the counter more than once, can increment, but only by 1, can increment, but only if it is action data and not coming from metadata, etc.). 

I am afraid that if we can't define PSA in a way to ensure reasonable compliance, it will become useless, because people will have no confidence that their programs could still work anywhere outside the original target and then they won't bother and claims that certain targets support PSA will also become hollow at best and misleading at worst. 

I am not sure we have a lot of good options here, given what I know about ASIC targets at least, but one way I am trying to explore is by limiting functionality to such a bare minimum that it can be implemented everywhere and fairly easily. That was the basis for my suggestion to get rid of byte counters (or accepting not-very-well-defined notion of packet length) and this is the basis for the comment above. 

The other important part of this whole endeavor is also the desire to ensure that a control plane, written to control a P4 program running on target A under PSA can control that same program on target B, meaning that the control plane interface to both standard P4 objects (tables) and PSA externs must be standardized. That's a very tall order, but without it PSA will probably not be very useful either. That's why I am a little worried that we started the effort without a good way to formalize the APIs -- comments and annotations do not seem to be enough, at least not to me.

Or, we can go back and scale down the goals. 

What do you think?